In [1]:
import tensorflow as tf
import numpy as np
from tensorflow import keras

In [2]:
tf.random.set_seed(42)

In [3]:
epochs = 200
batch_size = 128
verbose = 1
nb_classes = 10
n_hidden = 128
validation_split = 0.2
dropout = 0.3

In [4]:
mnist = keras.datasets.mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [5]:
print(X_train.shape, X_test.shape)

(60000, 28, 28) (10000, 28, 28)


In [6]:
reshaped = 784

X_train = X_train.reshape(60000, reshaped)
X_test = X_test.reshape(10000, reshaped)

In [7]:
X_train = X_train.astype(float)
X_test = X_test.astype(float)

In [8]:
X_train = X_train / 255
X_test = X_test / 255

In [9]:
y_train = tf.keras.utils.to_categorical(y_train, nb_classes)
y_test = tf.keras.utils.to_categorical(y_test, nb_classes)

In [10]:
model = tf.keras.models.Sequential()
model.add(keras.layers.Dense(n_hidden,
                             input_shape=(reshaped,),
                             activation='relu'))
model.add(keras.layers.Dropout(dropout))
model.add(keras.layers.Dense(n_hidden,
                             activation='relu'))
model.add(keras.layers.Dropout(dropout))
model.add(keras.layers.Dense(nb_classes,
                             activation='softmax'))

2022-03-16 16:44:43.720880: I tensorflow/core/platform/cpu_feature_guard.cc:145] This TensorFlow binary is optimized with Intel(R) MKL-DNN to use the following CPU instructions in performance critical operations:  SSE4.1 SSE4.2
To enable them in non-MKL-DNN operations, rebuild TensorFlow with the appropriate compiler flags.
2022-03-16 16:44:43.721412: I tensorflow/core/common_runtime/process_util.cc:115] Creating new thread pool with default inter op setting: 8. Tune using inter_op_parallelism_threads for best performance.


In [11]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dropout (Dropout)            (None, 128)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 128)               16512     
_________________________________________________________________
dropout_1 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 10)                1290      
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________________________________________________________________


In [12]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [13]:
model.fit(X_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=verbose,
          validation_split=validation_split)

Train on 48000 samples, validate on 12000 samples
Epoch 1/200
48000/48000 [==============================] - 4s 80us/sample - loss: 0.5261 - accuracy: 0.8392 - val_loss: 0.1893 - val_accuracy: 0.9456
Epoch 2/200
48000/48000 [==============================] - 3s 66us/sample - loss: 0.2336 - accuracy: 0.9306 - val_loss: 0.1347 - val_accuracy: 0.9601
Epoch 3/200
48000/48000 [==============================] - 3s 66us/sample - loss: 0.1772 - accuracy: 0.9473 - val_loss: 0.1153 - val_accuracy: 0.9654
Epoch 4/200
48000/48000 [==============================] - 3s 67us/sample - loss: 0.1500 - accuracy: 0.9537 - val_loss: 0.1028 - val_accuracy: 0.9707
Epoch 5/200
48000/48000 [==============================] - 3s 71us/sample - loss: 0.1290 - accuracy: 0.9603 - val_loss: 0.0989 - val_accuracy: 0.9711
Epoch 6/200
48000/48000 [==============================] - 4s 74us/sample - loss: 0.1144 - accuracy: 0.9645 - val_loss: 0.0943 - val_accuracy: 0.9734
Epoch 7/200
48000/48000 [=========================

Epoch 55/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0292 - accuracy: 0.9900 - val_loss: 0.0992 - val_accuracy: 0.9805
Epoch 56/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0308 - accuracy: 0.9902 - val_loss: 0.0976 - val_accuracy: 0.9798
Epoch 57/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0305 - accuracy: 0.9894 - val_loss: 0.0914 - val_accuracy: 0.9803
Epoch 58/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0323 - accuracy: 0.9895 - val_loss: 0.0983 - val_accuracy: 0.9805
Epoch 59/200
48000/48000 [==============================] - 3s 70us/sample - loss: 0.0287 - accuracy: 0.9907 - val_loss: 0.0939 - val_accuracy: 0.9804
Epoch 60/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0287 - accuracy: 0.9904 - val_loss: 0.1022 - val_accuracy: 0.9800
Epoch 61/200
48000/48000 [==============================] - 3s 70us/sample - loss: 0.0287 - ac

48000/48000 [==============================] - 3s 68us/sample - loss: 0.0235 - accuracy: 0.9928 - val_loss: 0.1089 - val_accuracy: 0.9793
Epoch 110/200
48000/48000 [==============================] - 3s 71us/sample - loss: 0.0195 - accuracy: 0.9937 - val_loss: 0.1178 - val_accuracy: 0.9803
Epoch 111/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0246 - accuracy: 0.9921 - val_loss: 0.1215 - val_accuracy: 0.9790
Epoch 112/200
48000/48000 [==============================] - 3s 68us/sample - loss: 0.0237 - accuracy: 0.9929 - val_loss: 0.1123 - val_accuracy: 0.9802
Epoch 113/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0219 - accuracy: 0.9930 - val_loss: 0.1091 - val_accuracy: 0.9801
Epoch 114/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0217 - accuracy: 0.9929 - val_loss: 0.1125 - val_accuracy: 0.9805
Epoch 115/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0207 - accuracy:

Epoch 163/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0199 - accuracy: 0.9935 - val_loss: 0.1197 - val_accuracy: 0.9808
Epoch 164/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0188 - accuracy: 0.9941 - val_loss: 0.1188 - val_accuracy: 0.9797
Epoch 165/200
48000/48000 [==============================] - 3s 68us/sample - loss: 0.0184 - accuracy: 0.9939 - val_loss: 0.1176 - val_accuracy: 0.9808
Epoch 166/200
48000/48000 [==============================] - 3s 67us/sample - loss: 0.0167 - accuracy: 0.9945 - val_loss: 0.1184 - val_accuracy: 0.9809
Epoch 167/200
48000/48000 [==============================] - 3s 69us/sample - loss: 0.0169 - accuracy: 0.9948 - val_loss: 0.1301 - val_accuracy: 0.9792
Epoch 168/200
48000/48000 [==============================] - 3s 72us/sample - loss: 0.0162 - accuracy: 0.9945 - val_loss: 0.1254 - val_accuracy: 0.9796
Epoch 169/200
48000/48000 [==============================] - 4s 77us/sample - loss: 0.01

In [14]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print('Test Accuracy', test_acc)

10000/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

Test Accuracy 0.9796


### Dropout 의 추가로 인해 정확도 상승
- Test Accuracy : from 0.9774 to 0.9796